File: src/main/java/com/magichamster/grocerysamurai/api/IBaseController.java
package com.magichamster.grocerysamurai.api;

import javax.ws.rs.*;
import javax.ws.rs.core.Response;

import com.magichamster.grocerysamurai.model.Identity;

public interface IBaseController<T extends Identity> {
	@POST
	Response add(T item);
	
	@PUT
	Response update(T item);
	
	@DELETE
	Response delete(@PathParam("id") int id);
	
	@GET
	T get(@PathParam("id") int id);
}

File: MagicHamster.GrocerySamurai/MagicHamster.GrocerySamurai.ServiceLayer/Controllers/IBaseController.cs
namespace MagicHamster.GrocerySamurai.ServiceLayer.Controllers
{
    using System.Threading.Tasks;
    using Microsoft.AspNetCore.Mvc;
    using Model.Common;

    public interface IBaseController<in T>
        where T : Entity
    {
        Task<IActionResult> GetAll(string userId = null, int? pageSize = 0);

        Task<IActionResult> Get(int? id);

        Task<IActionResult> Add(T record);

        Task<IActionResult> Update(T record);

        Task<IActionResult> Delete(int id);
    }
}